### DISCUSSION TITLE
"int" support for Tween.Custom()

### DISCUSSION THREAD

**Discussion by PhuNguyen182** [KyryloKuzyk/PrimeTween#87] [Ideas]
Hello, when using DOTween, I found incredible DOVirtual functions. Now I'm migrating to PrimeTween and I found Tween.Custom() do the same things as DoVirtual in DOTween. But why you support only float, Color, Vector, Quaternion and Rect but int? What if I need int overridden version of the Tween.Custom() in PrimeTween, because I use DoVirtual.Int() frequently in DOTween. I want to migrate thoroughly to PrimeTween due to the performance it brings. Thanks for your efforts.

**Comment by st4rdog** [comment]
If I'm understanding, you can cast like this for font size (int):

```c#
Tween.Custom(Text.fontSize, 110, duration: 1f, onValueChange: v => Text.fontSize = (int)v);
``` 

Sidenote, Scale should be Vector3 instead of Float...

**Comment by KyryloKuzyk** [comment]
Hey! As @st4rdog mentioned, it's possible to just round float to an int like this:
```csharp
Tween.Custom(text, startValue: 10f, endValue: 100f, duration: 1f, (target, val) => target.fontSize = Mathf.RoundToInt(val));
```

But this problem goes deeper than just rounding float to an int. I experimented with a version of Tween.Custom() that works with int and the problem with it much deeper than appears at first sight.

First, linearly interpolating an int is ambiguous in the sense that there is no mathematically correct behavior that will satisfy all use cases. Here are a few examples:
1. For a case when a game uses int coordinates for moving objects, what should be the correct way to do that? It seems like Mathf.RoundToInt() should work fine, right? Buuut, not quite.
For example, if you animate from 0 to 2 in 4 seconds, then the timeline will look like this (the "-" symbol represents one second):
"-1--2-"
The value of "1" is reported after 1 second, then 2 seconds pass before the value "2" is reported, then one second before the animation is finished.
It may be fine for some games, but other games may find this behavior limiting.

2. Another use case for int animation is a game that uses PrimeTween to animate the keyframes of a sprite animation.
In this case, you want each frame of the animation to be displayed for the same amount of time, so the approach of the previous example will not work.
If a sprite animation has 4 frames and you wish to play in 4 seconds, this is the expected timeline:
"1-2-3-4-"
How would you achieve that with Tween.Custom() if the int value is supported? You can only get a desired result with something [like this](https://github.com/KyryloKuzyk/PrimeTween/discussions/72#discussioncomment-9173467):
```csharp
tween = Tween.Custom(this, 0f, sprites.Length, settings, static (_this, val) => {
    int newIndex = (int)val;
    var _sprites = _this.sprites;
    if (newIndex < _sprites.Length && _this.curIndex != newIndex) {
        _this.curIndex = newIndex;
        _this.spriteRenderer.sprite = _sprites[newIndex];
    }
});
```

There are many other different examples of how one may want to animate an int value. But, unfortunately, I haven't found yet a solution that will satisfy all use cases for an int animation. It almost feels like animating an int value doesn't make any mathematical sense :)
I'll keep exploring this more and maybe I'll come with a good way to abstract this complexity away.
